buildtest/executors/base.py

```diff
@@ -45,15 +48,97 @@ def __repr__(self):
 
     def get(self, name):
         """Given the name of an executor return the executor for running 
-           a buildtest build.
+           a buildtest build, or get the default.
+        """
+        return self.executors.get(name) or self.executors.get("default")
+
+    def _choose_executor(self, builder):
+        """Choose executor is called at the onset of a run or dryrun. We
+           look at the builder metadata to determine if a default
+           is set for the executor, and fall back to the default.
+
+           :param builder: the builder with the loaded test configuration.
+           :type builder: buildtest.buildsystem.BuilderBase (or subclass).
+        """
+        executor = builder.metadata.get("executor", "default")
+
+        # The executor (or a default) must be define
+        if executor not in self.executors:
+            builder.logger.warning(
+                "executor %s is not defined in default.yml" % executor
+            )
+            sys.exit(1)
+
+        # Get the executor by name, and add the builder to it
+        executor = self.executors.get(executor)
+        executor.builder = builder
+        return executor
+
+    def dry_run(self, builder):
+        """A dry run typically includes all of the steps up to run
+
+           :param builder: the builder with the loaded test configuration.
+           :type builder: buildtest.buildsystem.BuilderBase (or subclass).
+        """
+        # Choose the executor based on the builder provided
+        executor = self._choose_executor(builder)
+
+        # Run each step defined for dry run
+        for step in executor.dryrun_steps:
+            if getattr(executor, step, None):
+                getattr(executor, step)()
+        return executor.result
+
+    def run(self, builder):
+        """Given a buildtest.buildsystem.BuildConfig (subclass) go through the
+           steps defined for the executor to run the build. This should
+           be instantiated by the subclass. For a simple script run, we expect a 
+           setup, build, and finish.
+
+           :param builder: the builder with the loaded test configuration.
+           :type builder: buildtest.buildsystem.BuilderBase (or subclass).
+        """
+        executor = self._choose_executor(builder)
+
+        # Run each step defined for dry run
+        for step in executor.steps:
+            if getattr(executor, step, None):
+                getattr(executor, step)()
+        return executor.result
```

Comment:
I am not understanding this ``run`` method from BuildExecutor that is returning value to variable ``result`` in https://github.com/researchapps/buildtest-framework/blob/add/executor-steps/buildtest/menu/build.py#L108 

I tested it locally it is returning ``None`` and getting the following error.

```
Traceback (most recent call last):
  File "/workspace/.pip-modules/bin/buildtest", line 11, in <module>
    load_entry_point('buildtest-framework', 'console_scripts', 'buildtest')()
  File "/workspace/buildtest-framework/buildtest/main.py", line 20, in main
    parser.parse_options()
  File "/workspace/buildtest-framework/buildtest/menu/__init__.py", line 80, in parse_options
    args.func(args)
  File "/workspace/buildtest-framework/buildtest/menu/build.py", line 111, in func_build_subcmd
    if result["RETURN_CODE"] == 0:
TypeError: 'NoneType' object is not subscriptable
```

Reply:
Can you please provide enough to reproduce this?

Reply:
okay reproduced with
```
$ buildtest build -c tests/testdir/slurm-hello.yml
```

Reply:
okay fixed the bug - one tiny typo! The step is called "run" and not "build" and I forgot to change it :)